<a href="https://colab.research.google.com/github/constantinouelenios-IB2AD0/constantinouelenios-IB2AD0-DataScience-GenAI-Submissions/blob/main/6_02_DNN_101.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://drive.google.com/uc?export=view&id=1xqQczl0FG-qtNA2_WQYuWePW9oU8irqJ)

# 6.02 Dense Neural Network (with PyTorch)
This will expand on our logistic regression example and take us through building our first neural network. If you haven't already, be sure to check (and if neccessary) switch to GPU processing by clicking Runtime > Change runtime type and selecting GPU. We can test this has worked with the following code:

In [ ]:
import torch

# Check for GPU availability
print("Num GPUs Available: ", torch.cuda.device_count())

Num GPUs Available:  1


Hopefully your code shows you have 1 GPU available! Next let's get some data. We'll start with another in-built dataset:

In [ ]:
# upload an in-built Python (OK semi-in-built) dataset
from sklearn.datasets import load_diabetes

import pandas as pd
import numpy as np

# import the data
data = load_diabetes()
data

{'data': array([[ 0.03807591,  0.05068012,  0.06169621, ..., -0.00259226,
          0.01990749, -0.01764613],
        [-0.00188202, -0.04464164, -0.05147406, ..., -0.03949338,
         -0.06833155, -0.09220405],
        [ 0.08529891,  0.05068012,  0.04445121, ..., -0.00259226,
          0.00286131, -0.02593034],
        ...,
        [ 0.04170844,  0.05068012, -0.01590626, ..., -0.01107952,
         -0.04688253,  0.01549073],
        [-0.04547248, -0.04464164,  0.03906215, ...,  0.02655962,
          0.04452873, -0.02593034],
        [-0.04547248, -0.04464164, -0.0730303 , ..., -0.03949338,
         -0.00422151,  0.00306441]]),
 'target': array([151.,  75., 141., 206., 135.,  97., 138.,  63., 110., 310., 101.,
         69., 179., 185., 118., 171., 166., 144.,  97., 168.,  68.,  49.,
         68., 245., 184., 202., 137.,  85., 131., 283., 129.,  59., 341.,
         87.,  65., 102., 265., 276., 252.,  90., 100.,  55.,  61.,  92.,
        259.,  53., 190., 142.,  75., 142., 155., 225.,  59

We are working on a regression problem, with "structured" data which has already been cleaned and normalised. We can skip the usual cleaning/engineering steps. However, we do need to get the data into PyTorch:

In [ ]:
# Convert data to PyTorch tensors
X = torch.tensor(data.data, dtype=torch.float32)
y = torch.tensor(data.target, dtype=torch.float32).reshape(-1, 1) # Reshape y to be a column vector

Now our data is stored in tensors we can do train/test splitting as before (in fact we can use sklearn as before):

In [ ]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

torch.Size([353, 10]) torch.Size([353, 1])
torch.Size([89, 10]) torch.Size([89, 1])


Now we can set up our batches for training. As we have a nice round 400 let's go with batches of 50 (8 batches in total). We'll also seperate the features and labels:

In [ ]:
from torch.utils.data import TensorDataset, DataLoader

# Create TensorDatasets and DataLoaders
train_dataset = TensorDataset(X_train, y_train)
train_loader = DataLoader(train_dataset, batch_size=50, shuffle=True)

test_dataset = TensorDataset(X_test, y_test)
test_loader = DataLoader(test_dataset, batch_size=50, shuffle=False)

Now its time to build our model. We'll keep it simple ... a model with an input layer of 10 features and then 2x _Dense_ (fully connected) layers each with 5 neurons and ReLU activation. Our output layer will be size=1 given this is a regression problem and we want a single value output per prediction.

This will be easier to understand if you have read through the logistic regression tutorial.

In [ ]:
import torch
import torch.nn as nn

# Define the model
class DiabetesModel(nn.Module):
    def __init__(self):
        super(DiabetesModel, self).__init__()
        # we'll set up the layers as a sequence using nn.Sequential
        self.layers = nn.Sequential(

            # first layer will be a linear layer that has 5x neurons
            # (5x sets of linear regression)
            # the layer takes the 10 features as input (i.e. 10, 5)
            nn.Linear(10, 5),

            nn.ReLU(), # ReLU activation

            # second linear layer again has 5 neurons
            # this time taking the input as the output of the last layer
            # (which had 5x neurons)
            nn.Linear(5, 5),

            nn.ReLU(), # ReLU again

            # last linear layer takes the output from the previous 5 neurons
            # this time its a single output with no activation
            # i.e. this is the predicitons (regression)
            nn.Linear(5, 1)
        )

    def forward(self, x):
        return self.layers(x) # pass the data through the layers

As before we need to create a model object, specify the loss (criterion) and an optimiser (which we cover next week):

In [ ]:
import torch.optim as optim

# Initialize the model, loss function, and optimizer
model = DiabetesModel()
criterion = nn.MSELoss() # MSE loss function
optimiser = optim.Adam(model.parameters(), lr=0.001)

Now we can train the model. Again, the logistic regression tutorial (6.01) may help you undertstand this:

In [ ]:
import torch

# Move model to GPU if available
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# Training loop (example - you'll likely want to add more epochs)
epochs = 1000 # 100 epochs

for epoch in range(epochs):
  # use the train_loader to pass the inputs (x) and targets (y)
  for inputs, targets in train_loader:
    # pass to the GPU (hopefully)
    inputs, targets = inputs.to(device), targets.to(device)

    # pass model to GPU as well
    model.to(device)

    model.train() # put the model object in train mode
    optimiser.zero_grad() # reset the gradiants
    outputs = model(inputs) # create outputs
    loss = criterion(outputs, targets) # compare with Y to get loss
    loss.backward() # backpropogate the loss (next week)
    optimiser.step() # # update the parameters based on this round of training

  # every 10 steps we will print out the current loss
    if (epoch+1) % 10 == 0: # modular arithmetic
        print(f'Epoch [{epoch+1}/{epochs}], Loss: {round(loss.item(), 4)}')

Epoch [10/1000], Loss: 18397.2188
Epoch [10/1000], Loss: 18960.2793
Epoch [10/1000], Loss: 19095.1152
Epoch [10/1000], Loss: 18774.9551
Epoch [10/1000], Loss: 16317.1494
Epoch [10/1000], Loss: 18642.7891
Epoch [10/1000], Loss: 18662.4238
Epoch [10/1000], Loss: 5852.3555
Epoch [20/1000], Loss: 15927.1299
Epoch [20/1000], Loss: 18052.4375
Epoch [20/1000], Loss: 13231.3506
Epoch [20/1000], Loss: 16848.5938
Epoch [20/1000], Loss: 15943.2119
Epoch [20/1000], Loss: 18242.4941
Epoch [20/1000], Loss: 14385.4951
Epoch [20/1000], Loss: 16608.6934
Epoch [30/1000], Loss: 15311.8643
Epoch [30/1000], Loss: 18082.7988
Epoch [30/1000], Loss: 15135.8145
Epoch [30/1000], Loss: 14106.8242
Epoch [30/1000], Loss: 10965.0752
Epoch [30/1000], Loss: 10892.9482
Epoch [30/1000], Loss: 12660.0283
Epoch [30/1000], Loss: 10401.6436
Epoch [40/1000], Loss: 9548.0723
Epoch [40/1000], Loss: 9737.2031
Epoch [40/1000], Loss: 13261.9805
Epoch [40/1000], Loss: 11411.7539
Epoch [40/1000], Loss: 13792.4482
Epoch [40/1000], 

We can see loss is significantly lower at the end than it was at the start. However, it is also bouncing around a little still which suggests the model needs more training (100 epochs is not a lot in deep learning terms). However, let's evaluate as before:

In [ ]:
# Evaluation (example)
model.eval() # testing mode
mse_values = [] # collect the MSE scores

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs) # predict the test data

        # Calculate Mean Squared Error
        mse = criterion(outputs, targets) # calcualte mse for the batch
        mse_values.append(mse.item()) # add to the list of MSE values

# Calculate and print the average MSE
avg_mse = np.mean(mse_values)
print(f"Average MSE on test set: {avg_mse}")

Average MSE on test set: 2855.1199951171875


MSE looks expected given training (no obvious sign of overfitting). However, we probably can get better results with tuning and more epochs.

Let's run the loop again a little differently to collect the predicted values (y_hat) and actuals (y) and add them to a dataset for comparions:

In [ ]:
# Evaluation
model.eval()
predictions = []
actuals = []

with torch.no_grad():
    for inputs, targets in test_loader:
        inputs, targets = inputs.to(device), targets.to(device)
        outputs = model(inputs)
        predictions.extend(outputs.cpu().numpy())
        actuals.extend(targets.cpu().numpy())

# Create DataFrame
results_df = pd.DataFrame({'Predicted': np.array(predictions).flatten(), 'Actual': np.array(actuals).flatten()})
results_df

Predicted  Actual
0   146.128174   219.0
1   175.910110    70.0
2   142.108093   202.0
3   294.398193   230.0
4   128.379059   111.0
..         ...     ...
84  116.957664   153.0
85   89.024422    98.0
86   77.176537    37.0
87   66.588280    63.0
88  153.455658   184.0

[89 rows x 2 columns]

Side-by-side, they don't look great. Can you improve them?

<br><br>

## EXERCISE #1
Try increasing the number of epochs to 1,000 (when the model is fairly well trained then the results printed for each 10x epochs will be fairly stable and not change much). Does this give better results?

<br><br>

## EXERCISE #2 (optional)
Try experimenting with the architecture (number of neurons and/or number of layers). Can we reach an optimal architecture?

### Results after Training with 1000 Epochs

After increasing the number of training epochs to 1000, the average Mean Squared Error (MSE) on the test set was approximately `2855.12`. This is comparable to the MSE obtained with 100 epochs, suggesting that simply training for more epochs did not lead to a significant improvement in model performance for this dataset. The model may have converged earlier, or further improvements might require architectural changes or hyperparameter tuning.

The side-by-side comparison of predicted versus actual values in the `results_df` indicates that while the model captures some general trends, there is still a noticeable discrepancy between the predictions and the true values. This reinforces the idea that further optimization beyond just increasing epochs is likely needed to achieve better predictive accuracy.

## Summary of Data Science and Modeling Process

This notebook demonstrates a basic end-to-end process for building and evaluating a Dense Neural Network for a regression problem using PyTorch, specifically predicting diabetes progression.

### 1. Data Loading and Understanding
*   **Dataset:** We used the in-built `load_diabetes()` dataset from `sklearn.datasets`. This dataset consists of **10 baseline variables** (age, sex, body mass index, average blood pressure, and six blood serum measurements) as features, and a quantitative measure of disease progression as the target.
*   **Nature of Data:** The data is 'structured' and already cleaned and normalized, simplifying the preprocessing steps.

### 2. Data Preparation for PyTorch
*   **Tensor Conversion:** The `numpy` arrays from `sklearn` were converted into PyTorch tensors using `torch.tensor()`.
*   **Target Reshaping:** The target variable `y` was reshaped to `(-1, 1)` to ensure it's a column vector, which is often required for PyTorch's loss functions.
*   **Train/Test Split:** The dataset was split into training (80%) and testing (20%) sets using `train_test_split` from `sklearn.model_selection`.
*   **DataLoaders:** `TensorDataset` and `DataLoader` from `torch.utils.data` were used to create iterable batches of data, enabling efficient training.

### 3. Model Architecture (Dense Neural Network)
*   **Model Class:** A `DiabetesModel` class was defined, inheriting from `nn.Module`.
*   **Layers:** The network consists of an input layer, two hidden dense (fully connected) layers, and an output layer:
    *   Input Layer: Takes 10 features as input.
    *   Hidden Layer 1: `nn.Linear(10, 5)` with 5 neurons and `nn.ReLU()` activation.
    *   Hidden Layer 2: `nn.Linear(5, 5)` with 5 neurons and `nn.ReLU()` activation.
    *   Output Layer: `nn.Linear(5, 1)` with 1 neuron (for regression) and no activation function.

### 4. Model Training
*   **Loss Function:** Mean Squared Error (`nn.MSELoss()`) was chosen, appropriate for regression tasks.
*   **Optimizer:** The Adam optimizer (`optim.Adam()`) was used to update model weights during training, with a learning rate of 0.001.
*   **Training Loop:** The model was trained over a specified number of `epochs` (initially 100, then increased to 1000). Each epoch iterates through the `train_loader`, performs forward pass, calculates loss, backpropagates gradients (`loss.backward()`), and updates weights (`optimiser.step()`).
*   **GPU Utilization:** The code checks for GPU availability and moves tensors and the model to the GPU if 'cuda' is available for faster training.

### 5. Model Evaluation
*   **Evaluation Mode:** The model was set to evaluation mode (`model.eval()`) to disable dropout and batch normalization updates.
*   **Metrics:** Average Mean Squared Error (MSE) on the test set was calculated to quantify performance.
*   **Prediction Comparison:** Predicted and actual values from the test set were collected and displayed in a `pandas.DataFrame` for visual inspection of the model's performance.